embulk-core/src/main/java/org/embulk/plugin/maven/MavenPluginSource.java

```diff
@@ -0,0 +1,168 @@
+
+        final MavenArtifactFinder mavenArtifactFinder;
+        try {
+            mavenArtifactFinder = MavenArtifactFinder.create(getLocalMavenRepository());
+        }
+        catch (MavenRepositoryNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Path jarPath;
+        try {
+            jarPath = mavenArtifactFinder.findMavenArtifactJar(
+                mavenPluginType.getGroup(),
+                "embulk-" + category + "-" + mavenPluginType.getName(),
+                mavenPluginType.getVersion());
+        }
+        catch (MavenArtifactNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Class pluginMainClass;
+        try (JarPluginLoader loader = JarPluginLoader.load(jarPath, pluginClassLoaderFactory)) {
+            pluginMainClass = loader.getPluginMainClass();
+        }
+        catch (InvalidJarPluginException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        if (!pluginInterface.isAssignableFrom(pluginMainClass)) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() + "\" is not a valid " + category + " plugin.");
+        }
+
+        final Object pluginMainObject;
+        try {
+            pluginMainObject = pluginMainClass.newInstance();
+        }
+        catch (InstantiationException ex) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() + "\" is not instantiatable.", ex);
+        }
+        catch (IllegalAccessException ex) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() +
+                "\" is not instantiatable due to access to constructor.", ex);
+        }
+        catch (ExceptionInInitializerError ex) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() +
+                "\" is not instantiatable due to exception in initialization.", ex);
+        }
+        catch (SecurityException ex) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() +
+                "\" is not instantiatable due to security manager.", ex);
+        }
+
+        try {
+            return pluginInterface.cast(pluginMainObject);
+        }
+        catch (ClassCastException ex) {
+            throw new PluginSourceNotMatchException(
+                "[FATAL/INTERNAL] Plugin class \"" + pluginMainClass.getName() + "\" is not " + category + " actually.",
+                ex);
+        }
+    }
+
+    private Path getLocalMavenRepository()
+            throws PluginSourceNotMatchException
+    {
+        String env;
+        try {
+            env = System.getenv("M2_REPO");
+        }
+        catch (NullPointerException | SecurityException ex) {
```

Comment:
Ah, exactly. But let me keep this not to forget in case we make it a kind of variables in the future. (E.g. from multiple environment variables, or else.)